pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -525,14 +616,114 @@ pub mod pallet {
+		/// message proof count for a particular message.
+		fn update_inbound_message_storages(
+			message: T::Message,
+			weight: &mut Weight,
+		) -> Result<(Proof, u32), DispatchError> {
+			match message.get_message_proof() {
+				None => {
+					let message_proof = message
+						.to_message_proof()
+						.get_message_proof()
+						.expect("message proof ensured by 'to_message_proof'");
+
+					match InboundMessages::<T>::try_mutate(message_proof, |storage_entry| {
+						match storage_entry {
+							None => {
+								*storage_entry =
+									Some((message, Self::get_expected_message_proof_count()));
+							}
+							Some((_, expected_proof_count)) => {
+								// We already have a message, in this case we should expect another
+								// set of message proofs.
+								expected_proof_count
+									.ensure_add_assign(Self::get_expected_message_proof_count())?;
+							}
+						};
+
+						Ok(())
+					}) {
+						Ok(_) => {}
+						Err(e) => return Err(e),
+					};
+
+					*weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+					Ok((
+						message_proof,
+						InboundMessageProofCount::<T>::get(message_proof),
+					))
+				}
+				Some(message_proof) => {
+					let message_proof_count =
+						match InboundMessageProofCount::<T>::try_mutate(message_proof, |count| {
+							count.ensure_add_assign(1)?;
+
+							Ok(*count)
+						}) {
+							Ok(r) => r,
+							Err(e) => return Err(e),
+						};
+
+					*weight = weight.saturating_add(T::DbWeight::get().writes(1));
+
+					Ok((message_proof, message_proof_count))
+				}
+			}
+		}
+
+		/// Ensures that the required number of message proofs is received
+		/// before sending the message to the `InboundMessageHandler` for
+		/// further processing.
 		fn process_inbound_message(
 			domain_address: DomainAddress,
 			message: T::Message,
 		) -> (DispatchResult, Weight) {
-			let weight = Weight::from_parts(0, T::Message::max_encoded_len() as u64)
+			let mut weight = Weight::from_parts(0, T::Message::max_encoded_len() as u64)
 				.saturating_add(LP_DEFENSIVE_WEIGHT);
 
+			let (message_proof, current_message_proof_count) =
+				match Self::update_inbound_message_storages(message, &mut weight) {
+					Ok(r) => r,
+					Err(e) => return (Err(e), weight),
+				};
+
+			weight = weight.saturating_add(T::DbWeight::get().reads(1));
+
+			let (message, total_expected_proof_count) =
+				match InboundMessages::<T>::get(message_proof) {
+					Some(r) => r,
+					None => {
+						// Not finding the message here is not a problem. We might have the correct
+						// message proof count but no actual message.
+						return (Ok(()), weight);
+					}
+				};
+
+			weight = weight.saturating_add(T::DbWeight::get().reads(1));
+
+			if current_message_proof_count < Self::get_expected_message_proof_count() {
+				return (Ok(()), weight);
+			}
+
+			if current_message_proof_count == total_expected_proof_count {
```

Comment:
@cdamian, just one important thing. How here do we know that the counts belong to different routers and not from the same router? I think this is where the concept of "vote" comes in

Reply:
That's the fun part, we do not...

Reply:
Not sure if we can use the domain_address though. If it comes via EVM it's most likely the same sender - the gateway router contract deployed on evm. Or maybe I'm mixing things up...

Reply:
It appears to me that we cannot rely on the `domain_address` based on our Slack conversation. We need to assume that in the future each domain has more than one configured router. So I would derive it "special origin" for the router, e.g. in case of Axelar `axelar_gateway_precompile::Pallet::<T>::GatewayContract` (see my [slack message](https://kflabs.slack.com/archives/C06FRHCRQL9/p1723044834860239?thread_ts=1723038229.117829&cid=C06FRHCRQL9)).